common/workunit/workflow.cpp

```diff
@@ -613,12 +808,888 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+        if(!cur.isActive())
+        {
+            cur.activate();
+            activateDependencies(cur);
+        }
+        //This break is for when a condition item is being activated.
+        //Out of its dependencies, only the condition expression should be made active.
+        //trueresult and falseresult should be left inactive
+        if(item.queryMode() == WFModeCondition)
+            break;
+    }
+}
+void WorkflowMachine::addToItemQueue(unsigned wfid)
+{
+    {
+        CriticalBlock thisBlock(queueCritSec);
+        wfItemQueue.push(wfid);
+    }
+    wfItemQueueSem.signal(1);
+}
+void WorkflowMachine::processSuccessors(CCloneWorkflowItem & item, Owned<IWorkflowDependencyIterator> iter, bool isLogical)
+{
+    //item will never be executed later, if it has been passed as an argument to this function
+    if(item.queryWfid() == startingWfid)
+    {
+        //update stop conditions
+        parentReached = true;
+#ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "Reached parent");
+#endif
+        //evaluate stop conditions.
+        //If failureClauseOld is false and the workflow has failed, then it needs to check whether there are any pending contingencies
+        checkIfDone();
+        return;
+    }
+    //net branches is used for when an item has failed
+    unsigned netBranches = 0U;
+    //if(abort == item.queryAbortFlag()) MORE: optional check that might increase speed, but may introduce race condition
+    for(iter->first();iter->isValid(); iter->next())
+    {
+        unsigned thisWfid = iter->query();
+        CCloneWorkflowItem & cur = queryWorkflowItem(thisWfid);
+        //item has an exception, implies that a dependency of cur has failed
+        if(item.queryException())
+        {
+            if(cur.queryException())
+            {
+                //only process the exception if cur is active
+                if(cur.isActive())
+                {
+                     cur.setAbortFlag(item.queryAbortFlag());
+                    addToItemQueue(thisWfid);
+                    netBranches++;
+                    if(reportFailureToFirstDependant)
+                        return;
+                }
+            }
+        }
+        //processDependentSuccessors made the current function call
+        else if(!isLogical)
+        {
+            unsigned numPred = cur.atomicDecNumDependencies();
+            if((numPred == 1) && cur.isActive())
+            {
+                cur.setAbortFlag(item.queryAbortFlag());
+                //ThreadSafe
+                addToItemQueue(thisWfid);
+            }
+        }
+        //processLogicalSuccessors made the current function call
+        else
+        {
+            bool itemIsReady = false;
+            if(!item.queryContingencyWithin())
+            {
+                //This means that if the item is already being executed and fails, it will cause a global abort, not a contingency failure
+                cur.setContingencyWithin(0);
```

Comment:
Is this condition required, give the code in line 1119?  That will only be executed if the item is ready - which may be significant.